master/internal/api_runs.go

```diff
@@ -508,3 +511,127 @@ func (a *apiServer) KillRuns(ctx context.Context, req *apiv1.KillRunsRequest,
+		ColumnExpr("r.experiment_id as exp_id").
+		ColumnExpr("((SELECT COUNT(*) FROM runs r WHERE e.id = r.experiment_id) > 1) as is_multitrial").
+		ColumnExpr("r.state IN (?) AS state", bun.In(model.StatesToStrings(model.TerminalStates))).
+		Join("LEFT JOIN experiments e ON r.experiment_id=e.id").
+		Join("JOIN projects p ON r.project_id = p.id").
+		Join("JOIN workspaces w ON p.workspace_id = w.id").
+		Where("r.project_id = ?", req.ProjectId)
+
+	if req.Filter == nil {
+		getQ = getQ.Where("r.id IN (?)", bun.In(req.RunIds))
+	} else {
+		getQ, err = filterRunQuery(getQ, req.Filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if getQ, err = experiment.AuthZProvider.Get().FilterExperimentsQuery(ctx, *curUser, nil, getQ,
+		[]rbacv1.PermissionType{
+			rbacv1.PermissionType_PERMISSION_TYPE_DELETE_EXPERIMENT,
+		}); err != nil {
+		return nil, err
+	}
+
+	err = getQ.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*apiv1.RunActionResult
+	visibleIDs := set.New[int32]()
+	var validIDs []int32
+	for _, check := range runChecks {
+		visibleIDs.Insert(check.ID)
+		switch {
+		case check.State == nil || !*check.State:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "Run is not in a terminal state.",
+				Id:    check.ID,
+			})
+		case check.Archived:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "Run is archived.",
+				Id:    check.ID,
+			})
+		default:
+			validIDs = append(validIDs, check.ID)
+		}
+	}
+	if req.Filter == nil {
+		for _, originalID := range req.RunIds {
+			if !visibleIDs.Contains(originalID) {
+				results = append(results, &apiv1.RunActionResult{
+					Error: fmt.Sprintf("Run with id '%d' not found in project with id '%d'", originalID, req.ProjectId),
+					Id:    originalID,
+				})
+			}
+		}
+	}
+	if len(validIDs) > 0 {
+		tx, err := db.Bun().BeginTx(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			txErr := tx.Rollback()
+			if txErr != nil && txErr != sql.ErrTxDone {
+				log.WithError(txErr).Error("error rolling back transaction in DeleteRuns")
+			}
+		}()
+		var acceptedIDs []int
+		if _, err = tx.NewDelete().Table("runs").
+			Where("runs.id IN (?)", bun.In(validIDs)).
+			Returning("runs.id").
+			Model(&acceptedIDs).
+			Exec(ctx); err != nil {
+			return nil, fmt.Errorf("delete runs: %w", err)
+		}
+
+		// delete run logs
+		if _, err = tx.NewDelete().Table("trial_logs").
+			Where("trial_logs.trial_id IN (?)", bun.In(acceptedIDs)).
+			Exec(ctx); err != nil {
+			return nil, fmt.Errorf("delete run logs: %w", err)
+		}
+
+		// delete task logs
+		trialTaskQuery := tx.NewSelect().TableExpr("tasks AS t").
+			ColumnExpr("t.task_id").
+			Join("JOIN run_id_task_id rt ON rt.task_id=t.task_id").
+			Where("rt.run_id IN (?)", bun.In(acceptedIDs))
+		if _, err = tx.NewDelete().Table("task_logs").
+			Where("task_logs.task_id IN (?)", trialTaskQuery).
+			Exec(ctx); err != nil {
+			return nil, fmt.Errorf("delete runs: %w", err)
+		}
```

Comment:
I guess these aren't being deleted by the cascade from trials.. we really need to relink tasks to runs and logs to runs instead. 

btw nothing has been in trial logs for a really long time. but I guess this is safe/necessary anyway, as long as the task delete cascades to task logs.

Reply:
Logs don't get cascade deleted. Added a test to check removing logs